src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -193,115 +132,262 @@ object Skat {
+          Row(key, size, q / (2 * sigmaSq), pval, fault)
+        } else {
+          Row(key, size, null, null, null)
+        }
+      }
+  }
 
-  /*
-  Davies algorithm takes the eigenvalues of (G * sqrt(W)).t * P_0 * G * sqrt(W)
-  We evaluate this matrix as 0.5 * (A.t * A - B.t * B) where
-  linear:   A = G * sqrt(W)              B = Q.t * G * sqrt(W)
-  logistic: A = sqrt(V) * G * sqrt(W)    B = inv(L) * X.t * V * G * sqrt(W)
-  Here X = Q * R and L * L.t = X.t * V * X
-  */
-  def runSkatPerKey(st: Array[SkatTuple], skatStatScaling: Double): SkatStat = {
+  def logisticSkat(keyGsWeightRdd: RDD[(Any, Iterable[(Vector[Double], Double)])],
+    y: DenseVector[Double], cov: DenseMatrix[Double], maxSize: Int, accuracy: Double, iterations: Int): RDD[Row] = {
+
+    val n = y.size
+    val k = cov.cols
+    val d = n - k
+
+    if (d < 1)
+      fatal(s"$n samples and $k ${ plural(k, "covariate") } including intercept implies $d degrees of freedom.")
+
+    val logRegM = new LogisticRegressionModel(cov, y).fit()
+    if (!logRegM.converged)
+      fatal("Failed to fit logistic regression null model (MLE with covariates only): " + (
+        if (logRegM.exploded)
+          s"exploded at Newton iteration ${ logRegM.nIter }"
+        else
+          "Newton iteration failed to converge"))
+
+    val mu = sigmoid(cov * logRegM.b)
+    val V = mu.map(x => x * (1 - x))
+    val VX = cov(::, *) :* V
+    val XtVX = cov.t * VX
+    XtVX.forceSymmetry()
+    var Cinv: DenseMatrix[Double] = null
+    try {
+      Cinv = inv(cholesky(XtVX))
+    } catch {
+      case _: MatrixSingularException => fatal("Singular matrix exception while computing Cholesky factor of X.t * V * X")
+      case _: NotConvergedException => fatal("Not converged exception while inverting Cholesky factor of X.t * V * X")
+    }
+    val res = y - mu
+
+    val sc = keyGsWeightRdd.sparkContext
+    val sqrtVBc = sc.broadcast(sqrt(V))
+    val resBc = sc.broadcast(res)
+    val CinvXtVBc = sc.broadcast(Cinv * VX.t)
+
+    def logisticTuple(x: Vector[Double], w: Double): SkatTuple = {
+      val xw = x * math.sqrt(w)
+      val sqrt_q = resBc.value dot xw      
+      SkatTuple(sqrt_q * sqrt_q, xw :* sqrtVBc.value , CinvXtVBc.value * xw)
+    }
+        
+    val maxEntries = math.max(maxSize * maxSize, maxEntriesForSmallN)
+
+    keyGsWeightRdd.map { case (key, vs) =>
+      val vsArray = vs.toArray
+      val size = vsArray.length
+      if (size <= maxSize) {
+        val skatTuples = vs.map((logisticTuple _).tupled).toArray
+        val (q, gramian) = if (size.toLong * n <= maxEntries) {
+          computeGramianSmallN(skatTuples)
+        } else {
+          computeGramianLargeN(skatTuples)
+        }
+        val (pval, fault) = computePval(q, gramian, accuracy, iterations)
+
+        // returning qstat = q / 2 to agree with skat R package convention
+        Row(key, size, q / 2, pval, fault)
+      } else {
+        Row(key, size, null, null, null)
+      }
+    }
```

Comment:
Seeing this again really make me think there should be a `LinearSkat` class and `LogisticSkat` class which both implement this trait:
```scala
trait SkatCore {
  // there must be a more satisfying name for this? is it rescaling something?
  def toTuples(x: Vector[Double], w: Double): SkatTuple
}
```
Which is really just `(Vector[Double], Double) => SkatTuple`.

The global fit should happen in their constructors / by-initializing-fields. With these classes in hand the `apply` method of `Skat` looks something like:

```scala
val fiddleVariant = if (logistic) new LogisticSkat() else new LinearSkat()

keyGsWeightRdd.map { case (key, vs) =>
  ...
    val skatTuples = vs.map(fiddleVariant.tupled).toArray
  ...
}
```